src/iter.rs

```diff
@@ -0,0 +1,685 @@
+                        }
+                    }
+                    TracePacket::GlobalTimestamp2 { ts } => {
+                        gts.upper = Some(ts);
+                        apply_gts(&gts, &mut self.current_baseline, &options);
+                    }
+
+                    packet => packets.push(packet),
+                },
+            }
+        }
+    }
+}
+
+impl<'a, R> Iterator for Timestamps<'a, R>
+where
+    R: Read,
+{
+    type Item = Result<TimestampedTracePackets, DecoderError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let trace = self.next_timestamped(self.options.clone());
+
+        match trace {
+            Err(DecoderErrorInt::Eof) => None,
+            Err(DecoderErrorInt::Io(io)) => Some(Err(DecoderError::Io(io))),
+            Err(DecoderErrorInt::MalformedPacket(m)) => Some(Err(DecoderError::MalformedPacket(m))),
+            Ok(trace) => Some(Ok(trace)),
+        }
+    }
+}
+
+fn calc_offset(ts: u64, prescaler: Option<LocalTimestampOptions>, freq: u32) -> chrono::Duration {
+    let prescale = match prescaler {
+        None | Some(LocalTimestampOptions::Enabled) => 1,
+        Some(LocalTimestampOptions::EnabledDiv4) => 4,
+        Some(LocalTimestampOptions::EnabledDiv16) => 16,
+        Some(LocalTimestampOptions::EnabledDiv64) => 64,
+        Some(LocalTimestampOptions::Disabled) => unreachable!(), // checked in `Timestamps::new`
+    };
+    let ticks = ts * prescale;
+    let seconds = (1_f64 / freq as f64) * ticks as f64;
+
+    chrono::Duration::nanoseconds((seconds * 1e9).ceil() as i64)
+}
+
+#[cfg(test)]
+mod timestamp_utils {
+    use super::*;
+
+    #[test]
+    fn gts() {
+        let gts = Gts {
+            lower: None,
+            upper: None,
+        };
+        assert_eq!(gts.merge(), None, "noop merge");
+
+        let mut gts = Gts {
+            lower: Some(42),
+            upper: Some(42),
+        };
+        assert_eq!(
+            gts.merge(),
+            Some((42 << Gts::GTS2_SHIFT) | 42),
+            "(42, 42) merge"
+        );
+
+        gts.replace_lower(0b1101011);
+        assert_eq!(
+            gts.merge(),
+            Some((42 << Gts::GTS2_SHIFT) | 0b1101011),
+            "replace whole merge"
+        );
+
+        let mut gts = Gts {
+            lower: Some(42),
+            upper: Some(42),
+        };
+        gts.replace_lower(1);
+        assert_eq!(
+            gts.merge(),
+            Some((42 << Gts::GTS2_SHIFT) | 43),
+            "replace partial merge"
+        );
+    }
+
+    #[test]
+    fn offset() {
+        assert_eq!(
+            calc_offset(1000, Some(LocalTimestampOptions::EnabledDiv4), 16_000_000),
+            chrono::Duration::microseconds(250),
+        );
+    }
+}
+
+#[cfg(test)]
+mod timestamps {
+    use super::{calc_offset, Gts};
+    use crate::*;
+    use std::ops::{Add, Sub};
+
+    const FREQ: u32 = 16_000_000;
+    static mut BASELINE: Option<chrono::DateTime<chrono::Utc>> = None;
```

Comment:
Used in all tests in this module. Safe as long as tests are not run in parallel.